// bkper/lib/model/Bkper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bkper = void 0;
const Book_1 = require("./Book");
const App_1 = require("./App");
const BookService = __importStar(require("../service/book-service"));
const UserService = __importStar(require("../service/user-service"));
const HttpApiRequest_1 = require("../service/HttpApiRequest");
const User_1 = require("./User");
/**
 * This is the main Entry Point of the [bkper-node](https://www.npmjs.com/package/bkper) library.
 *
 * @public
 */
class Bkper {
    /**
     * Gets the [[Book]] with the specified bookId from url param.
     *
     * @param id - The universal book id - The same bookId param of URL you access at app.bkper.com
     *
     * @returns The retrieved Book, for chaining
     */
    static getBook(id) {
        return __awaiter(this, void 0, void 0, function* () {
            let book = yield BookService.loadBook(id);
            return new Book_1.Book(book);
        });
    }
    /**
     * Gets the current logged [[User]].
     *
     * @returns The retrieved User, for chaining
     */
    static getUser() {
        return __awaiter(this, void 0, void 0, function* () {
            let user = yield UserService.getUser();
            return new User_1.User(user);
        });
    }
    /**
     * Sets the API [[Config]] object.
     *
     * @param config - The Config object
     */
    static setConfig(config) {
        HttpApiRequest_1.HttpApiRequest.config = config;
    }
    /**
     * Sets the API key to identify the agent.
     *
     * @param key - The API key
     *
     * @returns The defined [[App]] object
     *
     * @deprecated Use `setConfig()` instead
     */
    static setApiKey(key) {
        HttpApiRequest_1.HttpApiRequest.config.apiKeyProvider = () => __awaiter(this, void 0, void 0, function* () { return key; });
        return new App_1.App();
    }
    /**
     * Sets the provider of the valid OAuth2 access token
     *
     * @deprecated Use `setConfig()` instead
     */
    static setOAuthTokenProvider(oauthTokenProvider) {
        return __awaiter(this, void 0, void 0, function* () {
            HttpApiRequest_1.HttpApiRequest.config.oauthTokenProvider = oauthTokenProvider;
        });
    }
}
exports.Bkper = Bkper;
